### ISSUE TITLE
Add fix routine for L025

### ISSUE THREAD

**Issue by nolanbconaway** [sqlfluff/sqlfluff#404]
[This query](http://sqlfluff-online.herokuapp.com/fluffed?sql=H4sIACarKV8C_ytOzUlNLlHQUkgrys9V0CiGcA0VEosVEjUVikuTCktTiyoBBIbljSYAAAA%3D&dialect=postgres) has two problems:

```sh
$ echo 'select * from (select 1 as a) subquery' | sqlfluff lint -
== [stdin] FAIL
L:   1 | P:  31 | L011 | Implicit aliasing of table not allowed. Use explicit
                       | `AS` clause.
L:   1 | P:  31 | L025 | Alias 'subquery' is never used in SELECT statement.
```

It gets fixed as such 

```sh
$ echo 'select * from (select 1 as a) subquery' | sqlfluff fix - --rules L011,L010,L025                                
select * from (select 1 as a) as subquery
```

But if you lint the fixed sql, L025 remains!

```sh
$ echo 'select * from (select 1 as a) subquery' | sqlfluff fix - --rules L011,L010,L025 | sqlfluff lint -
== [stdin] FAIL
L:   1 | P:  34 | L025 | Alias 'subquery' is never used in SELECT statement.
```

The conjunction of these two errors shows that the "real" fix would be [this query](http://sqlfluff-online.herokuapp.com/fluffed?sql=H4sIAEGrKV8C_ytOzUlNLlHQUkgrys9V0CiGcA0VEosVEjV5uQDXINLGHwAAAA%3D%3D&dialect=postgres)

```sh
$ echo 'select * from (select 1 as a)' | sqlfluff lint -
# no violations!
```

Should fluff remove the implicit alias if it is never used?

**Comment by alanmcruickshank** [comment]
Great question.

So far I haven't gone as far as removing aliases if they're not used. I wondered whether it was too brutal, but actually, cases like this show that maybe it's more confusing if it's not there.

Would you be happy putting together a PR for this?

**Comment by nolanbconaway** [comment]
Lol, i knew the day would come in which I would need to learn the internals. I'll take this one!

**Comment by alanmcruickshank** [comment]
@nolanbconaway - feel free to ask any questions you need to as you go.

Some pointers:
- Each rule has an `_eval` function which is called on every segment of the query recursively.
- `_eval` returns a `LintResult` object which contains a selection of things about that linting operation.
- To make *corrections*, a `LintResult` should contain a single (or a list of) `LintFix` objects.
- A `LintFix` object has an `edit_type` of `delete`, `edit` or `create`. All have an `anchor`, which is a reference to the segment where the fix should act. Edit and create operations also have an `edit` property which specifies what edit to make.

One of the simpler rules which still implements fixing would be `Rule_L022` or even just `Rule_L001`. They might be instructive to examine and see how the fixing happens.

**Comment by nolanbconaway** [comment]
> @nolanbconaway - feel free to ask any questions you need to as you go.
> 
> Some pointers:
> 
> * Each rule has an `_eval` function which is called on every segment of the query recursively.
> * `_eval` returns a `LintResult` object which contains a selection of things about that linting operation.
> * To make _corrections_, a `LintResult` should contain a single (or a list of) `LintFix` objects.
> * A `LintFix` object has an `edit_type` of `delete`, `edit` or `create`. All have an `anchor`, which is a reference to the segment where the fix should act. Edit and create operations also have an `edit` property which specifies what edit to make.
> 
> One of the simpler rules which still implements fixing would be `Rule_L022` or even just `Rule_L001`. They might be instructive to examine and see how the fixing happens.

Thanks a bunch-- thats helpful context. I'll take a look and see what i see.

**Comment by nolanbconaway** [comment]
@alanmcruickshank are you thinking about this ticket much these days? I noticed you did some tagging, etc, recently. I am looking at it now and am not confident that _I_ will implement for 0.4.0 😬 

Perhaps a separate issue- but the main blocker for me (and perhaps other contributors new to the internals?) is that the the example simple rule you referenced (L022) is 150 lines and involves working with complicated objects + recursion. It's a lot to take in!

I have two suggestions to help new contributors wade in. I'm happy to open a separate issue for these matters, listing here only because they arose on this, my first attempt to work on `core` 😆 :

1. Document an example development flow, for _before_ you're ready to write unit tests. I don't know about you, but i usually work stuff out looking at my terminal before pytest gets involved. If I were editing `_eval` in L025, i'd have to spend time just figuring out the call stack that leads to printing in my terminal, let alone how to implement the feature I want.
2. Document an absolute minimal / unrealistic lint and fix rule example. This should describe what needs to be done to implement the [architecture](https://docs.sqlfluff.com/en/latest/architecture.html) to operate on SQl one-liners.


I don't have a solution in mind; but now i know how complicated fixing sql is!

**Comment by NiallRees** [comment]
I'd love to see this!

**Comment by barrywhart** [comment]
Mind if I take a look at this one, @nolanbconaway?

**Comment by nolanbconaway** [comment]
Please!

**Comment by barrywhart** [comment]
Great! If you're interested some time, I'd be happy to schedule an hour to pair program on an issue (Zoom screen share or something like that). Maybe that could be step 1 towards the "new contributor" docs you suggested. If you're interested, ping me on Slack and we'll set something up.

**Comment by nolanbconaway** [comment]
> Great! If you're interested some time, I'd be happy to schedule an hour to pair program on an issue (Zoom screen share or something like that). Maybe that could be step 1 towards the "new contributor" docs you suggested. If you're interested, ping me on Slack and we'll set something up.

That's really nice of you! I'll take you up on it when i get some more bandwidth

**Comment by barrywhart** [comment]
And if it helps, I'm usually pretty confused at the start of issues like this! The inheritance across several of the rules is a bit confusing. I also think it would be 💯 to add type hints to more of the code, especially the areas where new contributors are likely to land. Type hints can make it a lot easier to understand existing code. I think on this particular issue, I'll likely spend 80% of the time stepping through the existing code to understand it before making any useful changes.

**Comment by barrywhart** [comment]
As it turns out, the example query and fix is invalid syntax in every database I tested (MySQL, Postgres, T-SQL). Basically, "derived queries" require an alias. See issue #3545. For that issue, I'm adding an exception to the rule, so derived table aliases will never be removed, whether they are used or not.
